iree/turbine/kernel/wave/codegen/handlers.py

```diff
@@ -273,7 +267,109 @@ def handle_set_symbol(emitter: WaveEmitter, node: fx.Node):
 ###############################################################################
 
 
-def emit_mfma(m: int, n: int, k: int, acc: Value, values: list[Value]):
+def decompose_value(src: Value, bitsize: int) -> list[Value]:
```

Comment:
Going in line with what I said earlier, can we move this packing, unpacking and shuffles further up the compilation pipeline, so maybe in DecomposeReduceOps? The idea being maybe we introduce a pass that decomposes the mma into a reduce op and then we leverage the existing decomposition there and add the packing of shuffles. That would also simplify the code in the emitter.

Reply:
This is just a lowering quirk because `gpu.shuffle` AMDGPU lowering only supports 32bits (when the `gpu.shuffle` op itself supports arbitrary int/float/vector types). This really should be done on `gpu.shuffle` -> `llvm` lowering level. I don't think we will get anything useful by bringing it up in the stack and we should just fix lowering upstream.